File: Library/src/main/java/org/endeavourhealth/library/model/iml/ModelDocument.java
package org.endeavourhealth.library.model.iml;

import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import java.util.ArrayList;
import java.util.List;
import lombok.Getter;
import org.endeavourhealth.library.model.imq.QueryEntity;
import org.endeavourhealth.library.model.tripletree.TTContext;
import org.endeavourhealth.library.transforms.TTManager;

/**
 * A document containing various instance entities conforming to data model shapes of various kinds
 */
@JsonPropertyOrder({ "context", "query" })
public class ModelDocument {

  @Getter
  private TTContext context;

  @Getter
  private List<Entity> folder;

  @Getter
  private List<ConceptSet> conceptSet;

  private List<MapFunction> functionClause;

  @Getter
  private List<QueryEntity> query;

  public ModelDocument() {
    this.context = TTManager.getDefaultContext();
  }

  public ModelDocument setContext(TTContext context) {
    this.context = context;
    return this;
  }

  public ModelDocument setQuery(List<QueryEntity> query) {
    this.query = query;
    return this;
  }

  public ModelDocument addQuery(QueryEntity query) {
    if (this.query == null) this.query = new ArrayList<>();
    this.query.add(query);
    return this;
  }

  public ModelDocument setFolder(List<Entity> folder) {
    this.folder = folder;
    return this;
  }

  public ModelDocument addFolder(Entity folder) {
    if (this.folder == null) this.folder = new ArrayList<>();
    this.folder.add(folder);
    return this;
  }

  public ModelDocument setConceptSet(List<ConceptSet> conceptSet) {
    this.conceptSet = conceptSet;
    return this;
  }

  public ModelDocument addConceptSet(ConceptSet set) {
    if (this.conceptSet == null) this.conceptSet = new ArrayList<>();
    this.conceptSet.add(set);
    return this;
  }

  public List<MapFunction> getFunction() {
    return functionClause;
  }

  public ModelDocument setFunction(List<MapFunction> functionClause) {
    this.functionClause = functionClause;
    return this;
  }

  public ModelDocument addFunction(MapFunction functionClause) {
    if (this.functionClause == null) this.functionClause = new ArrayList<>();
    this.functionClause.add(functionClause);
    return this;
  }
}

File: src/interfaces/ModelDocument.ts
import z from "zod";

import { ConceptSetSchema } from "./ConceptSet";
import { EntitySchema } from "./Entity";
import { MapFunctionSchema } from "./MapFunction";
import { QueryEntitySchema } from "./QueryEntity";
import { TTContextSchema } from "./TTContext";

// export interface ModelDocument {
//   context?: TTContext;
//   query?: QueryEntity[];
//   folder?: Entity[];
//   conceptSet?: ConceptSet[];
//   function?: MapFunction[];
// }

export const ModelDocumentSchema = z.strictObject({
  context: TTContextSchema.optional(),
  query: z.array(QueryEntitySchema).prefault([]),
  folder: z.array(EntitySchema).prefault([]),
  conceptSet: z.array(ConceptSetSchema).prefault([]),
  function: z.array(MapFunctionSchema).prefault([])
});

export type ModelDocument = z.output<typeof ModelDocumentSchema>;
